napari/components/_dims/view.py

```diff
@@ -1,89 +1,230 @@
+
+        # ndims change listener
+        def update_ndims_listener(event):
+            self.update_ndims.emit()
+        self.dims.events.ndims.connect(update_ndims_listener)
+
+        # What to do with the ndims change events in terms of UI calls to the
+        # widget
+        self.update_ndims.connect(self._update_nsliders)
+
+    @property
+    def nsliders(self):
+        """Returns the number of sliders displayed
+
+        Returns
+        -------
+        nsliders: int
+            Number of sliders displayed
+        """
+        return len(self.sliders)
+
+    def _update_slider(self, slider_index: int):
+        """
+        Updates everything for a given slider
+
+        Parameters
+        ----------
+        slider_index : slider index (corresponds to axis index)
+        """
+        if slider_index >= self.nsliders:
+            return
+
+        slider = self.sliders[slider_index]
+
+        if slider is None:
+            return
 
-    def _axis_to_row(self, axis, max_dims):
-        message = f'axis {axis} out of bounds for {max_dims} dims'
+        if slider_index < self.dims.ndims:
 
-        if axis < 0:
-            axis = max_dims - axis
-            if axis < 0:
-                raise IndexError(message)
-        elif axis >= max_dims:
-            raise IndexError(message)
+            mode = self.dims.mode[slider_index]
+            if mode == DimsMode.POINT:
+                slider.collapse()
+                slider.setValue(self.dims.point[slider_index])
+            elif mode == DimsMode.INTERVAL:
+                slider.expand()
+                slider.setValues(self.dims.interval[slider_index])
+            slider_range = self.dims.range[slider_index]
+            if slider_range not in (None, (None, None, None)):
+                slider.setRange(slider_range)
 
-        if axis < 2:
-            raise ValueError('cannot convert y/x-axes to rows')
+    def _update_nsliders(self):
+        """
+        Updates the number of sliders based on the number of dimensions
+        """
+        self._set_nsliders(self.dims.ndims-2)
 
-        return axis - 2
+    def _set_nsliders(self, new_number_of_sliders):
+        """
+        Sets the number of sliders displayed
+        Parameters
+        ----------
+        new_number_of_sliders :
+        """
+        if self.nsliders < new_number_of_sliders:
+            self._create_sliders(new_number_of_sliders)
+        elif self.nsliders > new_number_of_sliders:
+            self._trim_sliders(new_number_of_sliders)
 
-    def update_slider(self, event):
-        """Updates a slider for the given axis or creates
-        it if it does not already exist.
+    def _create_sliders(self, number_of_sliders):
+        """
+        Creates sliders to match new number of dimensions
+        Parameters
+        ----------
+        number_of_sliders : new number of sliders
+        """
+        while number_of_sliders > self.nsliders:
```

Comment:
First, I would suggest 
```python
        while self.nsliders < number_of_sliders:
```

ie, put the variable that is changing in front. But, secondly, I don't like this open while loop when the variable updating is happening in a sub-function. It makes it very hard to assert that this is not an infinite loop. My bigger suggestion would be:

```suggestion
        for slider_num in range(self.nsliders, number_of_sliders):
```

Reply:
I agree - switched to `for` loop